app/addons/replication/components.react.jsx

```diff
@@ -0,0 +1,527 @@
+const ConfirmButton = Components.ConfirmButton;
+const PasswordModal = AuthComponents.PasswordModal;
+
+
+class ReplicationController extends React.Component {
+  constructor (props) {
+    super(props);
+    this.state = this.getStoreState();
+    this.submit = this.submit.bind(this);
+    this.clear = this.clear.bind(this);
+    this.showPasswordModal = this.showPasswordModal.bind(this);
+  }
+
+  getStoreState () {
+    return {
+      loading: store.isLoading(),
+      databases: store.getDatabases(),
+      authenticated: store.isAuthenticated(),
+      password: store.getPassword(),
+
+      // source fields
+      replicationSource: store.getReplicationSource(),
+      sourceDatabase: store.getSourceDatabase(),
+      localSourceDatabaseKnown: store.isLocalSourceDatabaseKnown(),
+      remoteSource: store.getRemoteSource(),
+
+      // target fields
+      replicationTarget: store.getReplicationTarget(),
+      targetDatabase: store.getTargetDatabase(),
+      localTargetDatabaseKnown: store.isLocalTargetDatabaseKnown(),
+      remoteTarget: store.getRemoteTarget(),
+
+      // other
+      passwordModalVisible: store.isPasswordModalVisible(),
+      replicationType: store.getReplicationType(),
+      replicationDocName: store.getReplicationDocName()
+    };
+  }
+
+  componentDidMount () {
+    store.on('change', this.onChange, this);
+  }
+
+  componentWillUnmount () {
+    store.off('change', this.onChange);
+  }
+
+  onChange () {
+    this.setState(this.getStoreState());
+  }
+
+  // the four local replication targets all show slightly different fields
+  getReplicationTargetRow () {
+    const { replicationTarget, remoteTarget, databases, targetDatabase } = this.state;
+    if (!replicationTarget) {
+      return null;
+    }
+    return (
+      <ReplicationTargetRow
+        remoteTarget={remoteTarget}
+        replicationTarget={replicationTarget}
+        databases={databases}
+        targetDatabase={targetDatabase}/>
+    );
+  }
+
+  clear (e) {
+    e.preventDefault();
+    Actions.clearReplicationForm();
+  }
+
+  showPasswordModal () {
+    const { replicationSource, replicationTarget } = this.state;
+
+    var hasLocalSourceOrTarget = (replicationSource === Constants.REPLICATION_SOURCE.LOCAL ||
+      replicationTarget === Constants.REPLICATION_TARGET.EXISTING_LOCAL_DATABASE ||
+      replicationTarget === Constants.REPLICATION_TARGET.NEW_LOCAL_DATABASE);
+
+    // if the user is authenticated, or if NEITHER the source nor target are local, just submit. The password
+    // modal isn't necessary
+    if (!hasLocalSourceOrTarget || this.state.authenticated) {
+      this.submit();
+      return;
+    }
+    AuthActions.showPasswordModal();
+  }
+
+  getUsername () {
+    return app.session.get('userCtx').name;
+  }
+
+  getAuthHeaders () {
+    const username = this.getUsername();
+    return {
+      'Authorization': 'Basic ' + base64.encode(username + ':' + this.state.password)
+    };
+  }
+
+  submit () {
```

Comment:
Could you break this `submit` up a bit into different functions depending on if its local etc. This is a pretty large function. Its a little tricky to follow.